File: backend/app/rag/vector_store.py
import os
import json
import logging
import numpy as np
from app.config import get_settings

logger = logging.getLogger(__name__)
settings = get_settings()

# Singleton — model loads once, reused on every request
_embedding_model = None

def _get_embedding_model():
    global _embedding_model
    if _embedding_model is None:
        from sentence_transformers import SentenceTransformer
        _embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
    return _embedding_model


class VectorStore:

    def __init__(self):
        self.store_type = settings.vector_store_type
        self.faiss_path = settings.faiss_index_path
        os.makedirs(self.faiss_path, exist_ok=True)
        self._metadata_store: dict[str, list[dict]] = {}
        self._load_metadata()

    async def generate_embeddings(self, texts: list[str]) -> list[list[float]]:
        import asyncio
        model = _get_embedding_model()
        loop = asyncio.get_event_loop()
        embeddings = await loop.run_in_executor(
            None,
            lambda: model.encode(texts, convert_to_numpy=True).tolist()
        )
        logger.info(f"Generated {len(embeddings)} embeddings locally")
        return embeddings

    async def add_documents(self, chunks: list, user_id: str):
        if not chunks:
            return
        texts = [chunk.text for chunk in chunks]
        embeddings = await self.generate_embeddings(texts)
        if self.store_type == "faiss":
            await self._add_to_faiss(embeddings, chunks, user_id)

    async def search(self, query: str, user_id: str, document_ids: list[str] = None, top_k: int = None) -> list[dict]:
        top_k = top_k or settings.top_k_retrieval
        query_embedding = await self.generate_embeddings([query])
        query_vector = query_embedding[0]
        if self.store_type == "faiss":
            return await self._search_faiss(query_vector, user_id, document_ids, top_k)
        return []

    async def _add_to_faiss(self, embeddings: list, chunks: list, user_id: str):
        import faiss

        dimension = len(embeddings[0])
        vectors = np.array(embeddings, dtype=np.float32)
        index_path = os.path.join(self.faiss_path, f"{user_id}.index")

        if os.path.exists(index_path):
            index = faiss.read_index(index_path)
            existing_metadata = self._metadata_store.get(user_id, [])
        else:
            index = faiss.IndexFlatIP(dimension)
            existing_metadata = []

        faiss.normalize_L2(vectors)
        index.add(vectors)

        for chunk in chunks:
            existing_metadata.append({
                "text": chunk.text,
                "page_number": chunk.page_number,
                "source_file": chunk.source_file,
                "document_id": chunk.document_id,
                "chunk_index": chunk.chunk_index,
            })

        self._metadata_store[user_id] = existing_metadata
        faiss.write_index(index, index_path)
        self._save_metadata()
        logger.info(f"Added {len(chunks)} vectors to FAISS for user {user_id}")

    async def _search_faiss(self, query_vector: list[float], user_id: str, document_ids: list[str] = None, top_k: int = 5) -> list[dict]:
        import faiss

        index_path = os.path.join(self.faiss_path, f"{user_id}.index")
        if not os.path.exists(index_path):
            return []

        index = faiss.read_index(index_path)
        metadata = self._metadata_store.get(user_id, [])

        if index.ntotal == 0:
            return []

        query_np = np.array([query_vector], dtype=np.float32)
        faiss.normalize_L2(query_np)

        search_k = min(top_k * 3 if document_ids else top_k, index.ntotal)
        scores, indices = index.search(query_np, search_k)

        results = []
        for score, idx in zip(scores[0], indices[0]):
            if idx == -1 or idx >= len(metadata):
                continue
            meta = metadata[idx]
            if document_ids and meta["document_id"] not in document_ids:
                continue
            results.append({"text": meta["text"], "score": float(score), "metadata": meta})
            if len(results) >= top_k:
                break

        return results

    async def delete_document_vectors(self, user_id: str, document_id: str):
        import faiss

        metadata = self._metadata_store.get(user_id, [])
        new_metadata = [m for m in metadata if m["document_id"] != document_id]

        if not new_metadata:
            index_path = os.path.join(self.faiss_path, f"{user_id}.index")
            if os.path.exists(index_path):
                os.remove(index_path)
            self._metadata_store[user_id] = []
            self._save_metadata()
            return

        texts = [m["text"] for m in new_metadata]
        embeddings = await self.generate_embeddings(texts)
        dimension = len(embeddings[0])
        vectors = np.array(embeddings, dtype=np.float32)
        faiss.normalize_L2(vectors)
        index = faiss.IndexFlatIP(dimension)
        index.add(vectors)
        index_path = os.path.join(self.faiss_path, f"{user_id}.index")
        faiss.write_index(index, index_path)
        self._metadata_store[user_id] = new_metadata
        self._save_metadata()

    def _save_metadata(self):
        meta_path = os.path.join(self.faiss_path, "metadata.json")
        with open(meta_path, "w") as f:
            json.dump(self._metadata_store, f)

    def _load_metadata(self):
        meta_path = os.path.join(self.faiss_path, "metadata.json")
        if os.path.exists(meta_path):
            with open(meta_path, "r") as f:
                self._metadata_store = json.load(f)


vector_store = VectorStore()

File: backend/src/rag/vectorStore.js
'use strict';
/**
 * Local vector store — cosine similarity over sentence-transformer embeddings.
 * Equivalent to Python's faiss.IndexFlatIP with L2 normalisation.
 *
 * Embeddings are generated via @xenova/transformers (all-MiniLM-L6-v2).
 * Data is persisted as binary (.vec) + JSON (.meta) files, one pair per user.
 */
const fs = require('fs');
const path = require('path');
const config = require('../config');

// Lazy-load the transformer pipeline (downloads model on first use)
let _pipeline = null;
async function getEmbeddingPipeline() {
  if (!_pipeline) {
    const { pipeline } = await import('@xenova/transformers');
    _pipeline = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2', {
      quantized: true,
    });
    console.log('[VectorStore] Embedding model loaded');
  }
  return _pipeline;
}

// ── Maths helpers ─────────────────────────────────────────────────────────────

function l2Normalize(vec) {
  const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0));
  return norm === 0 ? vec : vec.map(v => v / norm);
}

function dotProduct(a, b) {
  let s = 0;
  for (let i = 0; i < a.length; i++) s += a[i] * b[i];
  return s;
}

// ── Persistence helpers ───────────────────────────────────────────────────────

function vecPath(userId) {
  return path.join(path.resolve(config.vectorStorePath), `${userId}.vec`);
}
function metaPath(userId) {
  return path.join(path.resolve(config.vectorStorePath), `${userId}.meta.json`);
}

function loadIndex(userId) {
  const vp = vecPath(userId);
  const mp = metaPath(userId);
  if (!fs.existsSync(vp) || !fs.existsSync(mp)) return { vectors: [], metadata: [] };

  const meta = JSON.parse(fs.readFileSync(mp, 'utf8'));
  const buf = fs.readFileSync(vp);
  const dim = meta.length > 0 ? buf.byteLength / meta.length / 4 : 0;
  const vectors = [];
  for (let i = 0; i < meta.length; i++) {
    const vec = [];
    for (let j = 0; j < dim; j++) {
      vec.push(buf.readFloatLE((i * dim + j) * 4));
    }
    vectors.push(vec);
  }
  return { vectors, metadata: meta };
}

function saveIndex(userId, vectors, metadata) {
  if (vectors.length === 0) {
    [vecPath(userId), metaPath(userId)].forEach(p => fs.existsSync(p) && fs.unlinkSync(p));
    return;
  }
  const dim = vectors[0].length;
  const buf = Buffer.allocUnsafe(vectors.length * dim * 4);
  vectors.forEach((vec, i) => {
    vec.forEach((v, j) => buf.writeFloatLE(v, (i * dim + j) * 4));
  });
  fs.writeFileSync(vecPath(userId), buf);
  fs.writeFileSync(metaPath(userId), JSON.stringify(metadata));
}

// ── Public API ────────────────────────────────────────────────────────────────

async function generateEmbeddings(texts) {
  const pipe = await getEmbeddingPipeline();
  const embeddings = [];
  for (const text of texts) {
    const output = await pipe(text, { pooling: 'mean', normalize: false });
    const vec = Array.from(output.data);
    embeddings.push(l2Normalize(vec));
  }
  return embeddings;
}

async function addDocuments(chunks, userId) {
  if (!chunks.length) return;

  const texts = chunks.map(c => c.text);
  const newVectors = await generateEmbeddings(texts);

  const { vectors, metadata } = loadIndex(userId);

  for (let i = 0; i < chunks.length; i++) {
    const c = chunks[i];
    vectors.push(newVectors[i]);
    metadata.push({
      text: c.text,
      pageNumber: c.pageNumber,
      sourceFile: c.sourceFile,
      documentId: c.documentId,
      chunkIndex: c.chunkIndex,
    });
  }

  saveIndex(userId, vectors, metadata);
  console.log(`[VectorStore] Added ${chunks.length} vectors for user ${userId}`);
}

async function search(query, userId, documentIds = [], topK = null) {
  topK = topK || config.topKRetrieval;
  const { vectors, metadata } = loadIndex(userId);
  if (!vectors.length) return [];

  const [queryVec] = await generateEmbeddings([query]);

  const scores = vectors.map((vec, idx) => ({ idx, score: dotProduct(queryVec, vec) }));
  scores.sort((a, b) => b.score - a.score);

  const results = [];
  for (const { idx, score } of scores) {
    if (results.length >= topK) break;
    const meta = metadata[idx];
    if (documentIds.length && !documentIds.includes(meta.documentId)) continue;
    results.push({ text: meta.text, score, metadata: meta });
  }

  return results;
}

async function deleteDocumentVectors(userId, documentId) {
  const { vectors, metadata } = loadIndex(userId);
  const keep = metadata.map((m, i) => m.documentId !== documentId ? i : -1).filter(i => i !== -1);
  saveIndex(userId, keep.map(i => vectors[i]), keep.map(i => metadata[i]));
  console.log(`[VectorStore] Deleted vectors for document ${documentId}`);
}

module.exports = { addDocuments, search, deleteDocumentVectors, generateEmbeddings };
